js/id/actions/merge_remote_changes.js

```diff
@@ -0,0 +1,152 @@
+                    nodes.push.apply(nodes, c.b);
+                }
+                else if (_.isEqual(c.o, c.b)) {  // only changed locally
+                    nodes.push.apply(nodes, c.a);
+                }
+                else {       // changed both locally and remotely
+                    conflicts.push(t('merge_remote_changes.conflict.nodelist'));
+                    return;  // fail merge..
+                }
+            }
+        }
+
+        return target.update({nodes: nodes});
+    }
+
+    function mergeRemoteMembers(target) {
+        if (!target) return;
+
+        if (_.isEqual(target.members, remote.members)) {
+            return target;
+        }
+        if (option === 'force_remote') {
+            return target.update({members: remote.members});
+        }
+
+        conflicts.push(t('merge_remote_changes.conflict.memberlist'));
+        return;  // fail merge
+    }
+
+    function mergeRemoteTags(target) {
+        if (!target) return;
+
+        if (_.isEqual(target.tags, remote.tags)) {
+            return target;
+        }
+        if (option === 'force_remote') {
+            return target.update({tags: remote.tags});
+        }
+
+        var keys = _.reject(_.union(_.keys(base.tags), _.keys(remote.tags)), ignoreKey),
+            tags = _.clone(target.tags),
+            changed = false,
+            fail = false;
+
+        function ignoreKey(k) {
+            return k.indexOf('tiger:') === 0 || _.contains(iD.data.discarded, k);
+        }
+
+        for (var i = 0, imax = keys.length; i !== imax; i++) {
+            var k = keys[i];
+            if (remote.tags[k] !== base.tags[k]) {  // tag modified remotely..
+                if (target.tags[k] && target.tags[k] !== remote.tags[k]) {
+                    conflicts.push(t('merge_remote_changes.conflict.tags',
+                        { tag: k, local: target.tags[k], remote: remote.tags[k] }));
+                    fail = true;
+                }
+                else {
+                    tags[k] = remote.tags[k];
+                    changed = true;
+                }
+            }
+        }
+
+        return fail ? undefined : changed ? target.update({tags: tags}) : target;
+    }
+
+    var action = function(graph) {
+        var target = iD.Entity(local, {version: remote.version});
+
+        if (option === 'force_local') {
+            return graph.replace(target);
+        }
+
+        if (target.type === 'node') {
+            target = mergeLocation(target);
+        }
+        else if (target.type === 'way') {
+            graph.rebase(remoteGraph.childNodes(remote), [graph], false);
+            target = mergeRemoteChildren(target);
+        }
+        else if (target.type === 'relation') {
+            target = mergeRemoteMembers(target);
+        }
+
+        target = mergeRemoteTags(target);
+        return target ? graph.replace(target) : graph;
+    };
+
+    action.withOption = function(opt) {
```

Comment:
Only `safe` is used in practice, right? What's the purpose of `force_local` and `force_remote`?